toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+// a chroot to run repoquery.
+func getAllRepoData(repoUrls []string, workerTar, buildDir string) (namesToUrls map[string]string, err error) {
+	const (
+		leaveChrootOnDisk = false
+	)
+	timestamp.StartEvent("pull available package data from repos", nil)
+	defer timestamp.StopEvent(nil)
+
+	// Create a chroot to run repoquery in
+	// Create the directory 'buildDir' if it does not exist
+	exists, err := file.DirExists(buildDir)
+	if err != nil {
+		err = fmt.Errorf("failed to check if directory %s exists: %w", buildDir, err)
+		return nil, err
+	}
+	if !exists {
+		logger.Log.Infof("Pre-caching: Creating 1st time chroot directory %s", buildDir)
+		err = os.MkdirAll(buildDir, 0755)
+		if err != nil {
+			err = fmt.Errorf("failed to create directory %s: %w", buildDir, err)
+			return nil, err
+		}
+	}
+	queryChroot, err := createChroot(workerTar, buildDir, leaveChrootOnDisk)
+	if err != nil {
+		err = fmt.Errorf("failed to create chroot: %w", err)
+		return nil, err
+	}
+	defer queryChroot.Close(leaveChrootOnDisk)
+
+	namesToUrls = make(map[string]string)
+	for _, repoUrl := range repoUrls {
+		// Use the chroot to query each repo for the packages it contains
+		var packages []string
+		err = queryChroot.Run(func() (chrootErr error) {
+			packages, chrootErr = getRepoPackages(repoUrl)
+			return chrootErr
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// We will be searching by the name: "<name>-<version>.<distro>.<arch>", the results from the repoquery will be
+		// in the form of "<PARTIAL_URL>/<name>-<version>.<distro>.<arch>.rpm"
+		for _, pkgUrl := range packages {
+			name := path.Base(pkgUrl)
+			name = strings.TrimSuffix(name, ".rpm")
+
+			// We need to prepend the repoUrl to the partial URL to get the full URL
+			pkgUrl = fmt.Sprintf("%s/%s", repoUrl, pkgUrl)
+			namesToUrls[name] = pkgUrl
+		}
+	}
+	return
+}
+
+// createChroot creates a network-enabled chroot to run repoquery in. The caller is expected to call Close() on the
+// returned chroot unless an error is returned, in which case the chroot will be closed by this function.
+func createChroot(workerTar, chrootDir string, leaveChrootOnDisk bool) (queryChroot *safechroot.Chroot, err error) {
+	const (
+		dnfUtilsPackageName = "dnf-utils"
+		rootDir             = "/"
+	)
+	timestamp.StartEvent("creating repoquery chroot", nil)
+	defer timestamp.StopEvent(nil)
+	logger.Log.Info("Pre-cache: Creating chroot for repoquery")
+
+	queryChroot = safechroot.NewChroot(chrootDir, true)
+	err = queryChroot.Initialize(workerTar, nil, nil)
+	if err != nil {
+		err = fmt.Errorf("failed to initialize chroot: %w", err)
+		return
+	}
+
+	// We will need network to install the repoquery package
+	files := []safechroot.FileToCopy{
+		{Src: "/etc/resolv.conf", Dest: "/etc/resolv.conf"},
+	}
+	err = queryChroot.AddFiles(files...)
+	if err != nil {
+		err = fmt.Errorf("failed to add files to chroot: %w", err)
+		queryChroot.Close(false)
```

Comment:
To be on the safe side in case of any future changes adding more early returns, what do you think about conditionally calling `queryChroot.Close(false)` only once from a deferred call? Something like what `srpmpacker.go` does inside its `createChroot` function.

Reply:
Ah, I like that. Switched.